File: rs.py

def gf_add(x, y):
    return x ^ y

def gf_sub(x, y):
    return x ^ y # in binary galois field, subtraction is just the same as addition (since we mod 2)

def cl_mul(x,y):
    '''Bitwise carry-less multiplication on integers'''
    z = 0
    i = 0
    while (y>>i) > 0:
        if y & (1<<i):
            z ^= x<<i
        i += 1
    return z

def gf_mult_noLUT(x, y, prim=0):
    '''Multiplication in Galois Fields without using a precomputed look-up table (and thus it's slower)
    by using the standard carry-less multiplication + modular reduction using an irreducible prime polynomial'''

    ### Define bitwise carry-less operations as inner functions ###
    def cl_mult(x,y):
        '''Bitwise carry-less multiplication on integers'''
        z = 0
        i = 0
        while (y>>i) > 0:
            if y & (1<<i):
                z ^= x<<i
            i += 1
        return z

    def bit_length(n):
        '''Compute the position of the most significant bit (1) of an integer. Equivalent to int.bit_length()'''
        bits = 0
        while n >> bits: bits += 1
        return bits

    def cl_div(dividend, divisor=None):
        '''Bitwise carry-less long division on integers and returns the remainder'''
        # Compute the position of the most significant bit for each integers
        dl1 = bit_length(dividend)
        dl2 = bit_length(divisor)
        # If the dividend is smaller than the divisor, just exit
        if dl1 < dl2:
            return dividend
        # Else, align the most significant 1 of the divisor to the most significant 1 of the dividend (by shifting the divisor)
        for i in range(dl1-dl2,-1,-1):
            # Check that the dividend is divisible (useless for the first iteration but important for the next ones)
            if dividend & (1 << i+dl2-1):
                # If divisible, then shift the divisor to align the most significant bits and XOR (carry-less subtraction)
                dividend ^= divisor << i
        return dividend
    
    ### Main GF multiplication routine ###

    # Multiply the gf numbers
    result = cl_mult(x,y)
    # Then do a modular reduction (ie, remainder from the division) with an irreducible primitive polynomial so that it stays inside GF bounds
    if prim > 0:
        result = cl_div(result, prim)

    return result


def gf_mult_noLUT(x, y, prim=0, field_charac_full=256, carryless=True):
    '''Galois Field integer multiplication using Russian Peasant Multiplication algorithm (faster than the standard multiplication + modular reduction).
    If prim is 0 and carryless=False, then the function produces the result for a standard integers multiplication (no carry-less arithmetics nor modular reduction).'''
    r = 0
    while y: # while y is above 0
        if y & 1: r = r ^ x if carryless else r + x # y is odd, then add the corresponding x to r (the sum of all x's corresponding to odd y's will give the final product). Note that since we're in GF(2), the addition is in fact an XOR (very important because in GF(2) the multiplication and additions are carry-less, thus it changes the result!).
        y = y >> 1 # equivalent to y // 2
        x = x << 1 # equivalent to x*2
        if prim > 0 and x & field_charac_full: x = x ^ prim # GF modulo: if x >= 256 then apply modular reduction using the primitive polynomial (we just subtract, but since the primitive number can be above 256 then we directly XOR).

    return r

gf_exp = [0] * 512 # Create list of 512 elements. In Python 2.6+, consider using bytearray
gf_log = [0] * 256

def init_tables(prim=0x11d):
    '''Precompute the logarithm and anti-log tables for faster computation later, using the provided primitive polynomial.'''
    # prim is the primitive (binary) polynomial. Since it's a polynomial in the binary sense,
    # it's only in fact a single galois field value between 0 and 255, and not a list of gf values.
    global gf_exp, gf_log
    gf_exp = [0] * 512 # anti-log (exponential) table
    gf_log = [0] * 256 # log table
    # For each possible value in the galois field 2^8, we will pre-compute the logarithm and anti-logarithm (exponential) of this value
    x = 1
    for i in range(0, 255):
        gf_exp[i] = x # compute anti-log for this value and store it in a table
        gf_log[x] = i # compute log at the same time
        x = gf_mult_noLUT(x, 3, prim)

        # If you use only generator==2 or a power of 2, you can use the following which is faster than gf_mult_noLUT():
        #x <<= 1 # multiply by 2 (change 1 by another number y to multiply by a power of 2^y)
        #if x & 0x100: # similar to x >= 256, but a lot faster (because 0x100 == 256)
            #x ^= prim # substract the primary polynomial to the current value (instead of 255, so that we get a unique set made of coprime numbers), this is the core of the tables generation

    # Optimization: double the size of the anti-log table so that we don't need to mod 255 to
    # stay inside the bounds (because we will mainly use this table for the multiplication of two GF numbers, no more).
    for i in range(255, 512):
        gf_exp[i] = gf_exp[i - 255]
    return [gf_log, gf_exp]

def gf_mul(x,y):
    if x==0 or y==0:
        return 0
    #print (f">gf mul: {gf_exp[gf_log[x] + gf_log[y]]}")
    #print(f"x: {gf_log[x]} y: {gf_log[y]} result: {gf_exp[gf_log[x] + gf_log[y]]}")
    return gf_exp[gf_log[x] + gf_log[y]] # should be gf_exp[(gf_log[x]+gf_log[y])%255] if gf_exp wasn't oversized

def gf_div(x,y):
    if y==0:
        raise ZeroDivisionError()
    if x==0:
        return 0
    return gf_exp[(gf_log[x] + 255 - gf_log[y]) % 255]


def gf_pow(x, power):
    print(f">gf pow: {gf_exp[(gf_log[x] * power) % 255]} with x={x}, power={power}")
    return gf_exp[(gf_log[x] * power) % 255]

def gf_inverse(x):
    return gf_exp[255 - gf_log[x]] # gf_inverse(x) == gf_div(1, x)

def gf_poly_scale(p,x):
    r = [0] * len(p)
    for i in range(0, len(p)):
        r[i] = gf_mul(p[i], x)
    return r

def gf_poly_add(p,q):
    r = [0] * max(len(p),len(q))
    for i in range(0,len(p)):
        r[i+len(r)-len(p)] = p[i]
    for i in range(0,len(q)):
        r[i+len(r)-len(q)] ^= q[i]
    return r

def gf_poly_mul(p,q):
    '''Multiply two polynomials, inside Galois Field'''
    # Pre-allocate the result array
    r = [0] * (len(p)+len(q)-1)
    # Compute the polynomial multiplication (just like the outer product of two vectors,
    # we multiply each coefficients of p with all coefficients of q)
    for j in range(0, len(q)):
        for i in range(0, len(p)):
            #print(f"before===> {r[i+j]}")
            #print(f"before===> {r}")
            #print(f"xor({r[i+j]}, {gf_mul(p[i], q[j])}) = {r[i+j] ^ gf_mul(p[i], q[j]) }")
            r[i+j] ^= gf_mul(p[i], q[j]) # equivalent to: r[i + j] = gf_add(r[i+j], gf_mul(p[i], q[j]))
                                                         # -- you can see it's your usual polynomial multiplication
            #print(f"after===> {r[i+j]}")                                                         
            #print(f"after===> {r}")
            #print()
    return r

def gf_poly_eval(poly, x):
    '''Evaluates a polynomial in GF(2^p) given the value for x. This is based on Horner's scheme for maximum efficiency.'''
    y = poly[0]
    for i in range(1, len(poly)):
        y = gf_mul(y, x) ^ poly[i]
    return y

def rs_generator_poly(nsym):
    '''Generate an irreducible generator polynomial (necessary to encode a message into Reed-Solomon)'''
    g = [1]
    for i in range(0, nsym):
        print(f"tmp: {[1, gf_pow(3, i+1)]}")
        g = gf_poly_mul(g, [1, gf_pow(3, i+1)])
    return g

def gf_poly_div(dividend, divisor):
    '''Fast polynomial division by using Extended Synthetic Division and optimized for GF(2^p) computations
    (doesn't work with standard polynomials outside of this galois field, see the Wikipedia article for generic algorithm).'''
    # CAUTION: this function expects polynomials to follow the opposite convention at decoding:
    # the terms must go from the biggest to lowest degree (while most other functions here expect
    # a list from lowest to biggest degree). eg: 1 + 2x + 5x^2 = [5, 2, 1], NOT [1, 2, 5]

    msg_out = list(dividend) # Copy the dividend
    #normalizer = divisor[0] # precomputing for performance
    for i in range(0, len(dividend) - (len(divisor)-1)):
        #msg_out[i] /= normalizer # for general polynomial division (when polynomials are non-monic), the usual way of using
                                  # synthetic division is to divide the divisor g(x) with its leading coefficient, but not needed here.
        coef = msg_out[i] # precaching
        if coef != 0: # log(0) is undefined, so we need to avoid that case explicitly (and it's also a good optimization).
            for j in range(1, len(divisor)): # in synthetic division, we always skip the first coefficient of the divisior,
                                              # because it's only used to normalize the dividend coefficient
                if divisor[j] != 0: # log(0) is undefined
                    msg_out[i + j] ^= gf_mul(divisor[j], coef) # equivalent to the more mathematically correct
                                                               # (but xoring directly is faster): msg_out[i + j] += -divisor[j] * coef

    # The resulting msg_out contains both the quotient and the remainder, the remainder being the size of the divisor
    # (the remainder has necessarily the same degree as the divisor -- not length but degree == length-1 -- since it's
    # what we couldn't divide from the dividend), so we compute the index where this separation is, and return the quotient and remainder.
    separator = -(len(divisor)-1)
    print(f"gf_poly_div: {dividend}, {divisor}, {msg_out}")
    return msg_out[:separator], msg_out[separator:] # return quotient, remainder.

def rs_encode_msgt(msg_in, nsym):
    '''Reed-Solomon main encoding function'''
    gen = rs_generator_poly(nsym)

    # Pad the message, then divide it by the irreducible generator polynomial
    _, remainder = gf_poly_div(msg_in + [0] * (len(gen)-1), gen)
    # The remainder is our RS code! Just append it to our original message to get our full codeword (this represents a polynomial of max 256 terms)
    msg_out = msg_in + remainder
    # Return the codeword
    return msg_out

def rs_encode_msg(msg_in, nsym):
    '''Reed-Solomon main encoding function, using polynomial division (algorithm Extended Synthetic Division)'''
    if (len(msg_in) + nsym) > 255: raise ValueError("Message is too long (%i when max is 255)" % (len(msg_in)+nsym))
    gen = rs_generator_poly(nsym)
    
    print("==========================")
    print(f"message: {msg_in}")
    print(f"nsym: {nsym}")
    print(f"generator poly:        {gen}")
    print(f"generator poly binary: ",end="")
    for i in gen:
        print(f'{i:>08b}',end=" ")
    print("\n==========================")

    # Init msg_out with the values inside msg_in and pad with len(gen)-1 bytes (which is the number of ecc symbols).
    msg_out = [0] * (len(msg_in) + len(gen)-1)
    # Initializing the Synthetic Division with the dividend (= input message polynomial)
    msg_out[:len(msg_in)] = msg_in

    print(f"msg_out: {msg_out}")

    # Synthetic division main loop
    for i in range(len(msg_in)):
        # Note that it's msg_out here, not msg_in. Thus, we reuse the updated value at each iteration
        # (this is how Synthetic Division works: instead of storing in a temporary register the intermediate values,
        # we directly commit them to the output).
        coef = msg_out[i]

        # log(0) is undefined, so we need to manually check for this case. There's no need to check
        # the divisor here because we know it can't be 0 since we generated it.
        if coef != 0:
            # in synthetic division, we always skip the first coefficient of the divisior, because it's only used to normalize the dividend coefficient (which is here useless since the divisor, the generator polynomial, is always monic)
            for j in range(1, len(gen)):
                #print(msg_out[i+j])
                #print(gf_mul(gen[j], coef))
                #print(f"genj: {gen[j]} coef: {coef}")
                msg_out[i+j] ^= gf_mul(gen[j], coef) # equivalent to msg_out[i+j] += gf_mul(gen[j], coef)
                #print(i, j)
                #print(msg_out)

    # At this point, the Extended Synthetic Divison is done, msg_out contains the quotient in msg_out[:len(msg_in)]
    # and the remainder in msg_out[len(msg_in):]. Here for RS encoding, we don't need the quotient but only the remainder
    # (which represents the RS code), so we can just overwrite the quotient with the input message, so that we get
    # our complete codeword composed of the message + code.
    print(msg_out)
    msg_out[:len(msg_in)] = msg_in

    return msg_out

prim = 0x11b
n = 32 # set the size you want, it must be > k, the remaining n-k symbols will be the ECC code (more is better)
k = 25 # k = len(message)
message = "Bezdratove a mobilni site" # input message

# Initializing the log/antilog tables
init_tables(prim)
#print(gf_log)
#print(gf_exp)

# Encoding the input message
mesecc = rs_encode_msg([ord(x) for x in message], n-k)

print("\nresult:")
print(mesecc)
for i in mesecc:
    #print(f'{i:>2} in binary is {i:>08b}')
    print(f'{i:>08b}',end="")

File: rs.cc
/**
 * @file rs.cc
 * @author Michal Findra, xfindr00
 * @brief Reed-Salomon code operations
 * @date 12.12.2022
 */

// other libs
#include "rs.hh"

#include <algorithm>
#include <iostream>
#include <iterator>
#include <string>
#include <vector>

BMS_RS::BMS_RS() {
}

std::vector<int> BMS_RS::rs_generator_poly(int nsym) {
    // Generate an irreducible generator polynomial (necessary to encode a message into Reed-Solomon)
    std::vector<int> g;
    g.push_back(1);
    for (int i = 1; i <= nsym; i++) {
        std::vector<int> tmp;
        tmp.push_back(1);
        tmp.push_back(bms_gf.gf_pow(3, i));
        g = bms_gf.gf_poly_mul(g, tmp);
    }
    return g;
}

std::vector<int> BMS_RS::rs_encode_msg(const std::vector<int>& msg_in, int nsym) {
    if (msg_in.size() + nsym > 255) {
        std::cerr << "ERROR: message too long" << std::endl;
        exit(EXIT_FAILURE);
    }

    std::vector<int> gen = rs_generator_poly(nsym);
    std::vector<int> msg_out(msg_in.size() + gen.size() - 1);

    // initialize the first part of the output message with the input message
    for (int i = 0; i < msg_in.size(); i++) {
        msg_out[i] = msg_in[i];
    }

    for (int i = 0; i < msg_in.size(); i++) {
        int coef = msg_out[i];

        if (coef != 0) {
            for (int j = 1; j < gen.size(); j++) {
                msg_out[i + j] ^= bms_gf.gf_mul(gen[j], coef);
            }
        }
    }
    return msg_out;
}

std::vector<int> BMS_RS::rs_calc_syndromes(const std::vector<int>& msg, int nsym) {
    std::vector<int> synd(nsym);
    for (int i = 0; i < nsym; ++i) {
        synd[i] = bms_gf.gf_poly_eval(msg, bms_gf.gf_pow(3, i + 1));
    }

    // Note the "[0] +" : we add a 0 coefficient for the lowest degree (the constant). This effectively shifts the syndrome, and will shift every computations depending on the syndromes (such as the errors locator polynomial, errors evaluator polynomial, etc. but not the errors positions).
    // This is not necessary, you can adapt subsequent computations to start from 0 instead of skipping the first iteration (ie, the often seen range(1, n-k+1)),
    synd.insert(synd.begin(), 0);

    return synd;
}

bool BMS_RS::rs_check(const std::vector<int>& msg, int nsym) {
    std::vector<int> synd = rs_calc_syndromes(msg, nsym);
    auto it = std::max_element(synd.begin(), synd.end());
    return (*it == 0);
}

std::vector<int> BMS_RS::rs_find_errata_locator(const std::vector<int>& e_pos) {
    // Compute the erasures/errors/errata locator polynomial from the erasures/errors/errata positions
    // (the positions must be relative to the x coefficient, eg: "hello worldxxxxxxxxx" is tampered to "h_ll_ worldxxxxxxxxx"
    // with xxxxxxxxx being the ecc of length n-k=9, here the string positions are [1, 4], but the coefficients are reversed
    // since the ecc characters are placed as the first coefficients of the polynomial, thus the coefficients of the
    // erased characters are n-1 - [1, 4] = [18, 15] = erasures_loc to be specified as an argument.

    std::vector<int> e_loc{1};  // just to init because we will multiply, so it must be 1 so that the multiplication starts correctly without nulling any term
    // erasures_loc = product(1 - x*alpha**i) for i in erasures_pos and where alpha is the alpha chosen to evaluate polynomials.
    for (const int i : e_pos) {
        e_loc = bms_gf.gf_poly_mul(e_loc, bms_gf.gf_poly_add({1}, {bms_gf.gf_pow(3, i), 0}));
    }

    return e_loc;
}

std::vector<int> BMS_RS::rs_find_error_evaluator(const std::vector<int>& synd, const std::vector<int>& err_loc, int nsym) {
    // Omega(x) = [ Synd(x) * Error_loc(x) ] mod x^(n-k+1)

    std::vector<int> v{1};
    std::vector<int> tmp(nsym + 1, 0);
    v.insert(v.end(), tmp.begin(), tmp.end());

    std::vector<int> remainder = bms_gf.gf_poly_div(bms_gf.gf_poly_mul(synd, err_loc), v);
    return remainder;
}

std::vector<int> BMS_RS::rs_find_errors(std::vector<int>& err_loc, int nmess) {
    int errs = err_loc.size() - 1;
    std::vector<int> err_pos;
    for (int i = 0; i < nmess; ++i) {
        if (bms_gf.gf_poly_eval(err_loc, bms_gf.gf_pow(3, i)) == 0) {
            err_pos.push_back(nmess - 1 - i);
        }
    }
    // Sanity check: the number of errors/errata positions found should be exactly the same as the length of the errata locator polynomial
    if (err_pos.size() != errs) {
        // couldn't find error locations
        std::cerr << "ERROR: too many errors" << std::endl;
        exit(EXIT_FAILURE);
    }
    return err_pos;
}

std::vector<int> BMS_RS::rs_find_error_locator(const std::vector<int>& synd, int nsym, int erase_count) {
    std::vector<int> err_loc{1};
    std::vector<int> old_loc{1};

    int synd_shift = synd.size() - nsym;
    for (int i = 0; i < nsym - erase_count; i++) {
        int K = i + synd_shift;  //(erase_loc.empty()) ? (i + synd_shift) : (erase_count + i + synd_shift);
        int delta = synd[K];
        for (int j = 1; j < err_loc.size(); j++) {
            delta ^= bms_gf.gf_mul(err_loc.at(err_loc.size() - (j + 1)), synd[K - j]);
        }

        old_loc.push_back(0);

        if (delta != 0) {
            if (old_loc.size() > err_loc.size()) {
                std::vector<int> new_loc = bms_gf.gf_poly_scale(old_loc, delta);
                old_loc = bms_gf.gf_poly_scale(err_loc, bms_gf.gf_inverse(delta));
                err_loc = new_loc;
            }
            err_loc = bms_gf.gf_poly_add(err_loc, bms_gf.gf_poly_scale(old_loc, delta));
        }
    }

    int count = 0;
    for (int i = 0; i < err_loc.size(); i++) {
        if (err_loc[i] == 0) {
            count++;
        } else {
            break;
        }
    }

    err_loc.erase(err_loc.begin(), err_loc.begin() + count);

    int errs = err_loc.size() - 1;
    if ((errs)*2 > nsym) {
        std::cerr << "ERROR: too many errors" << std::endl;
        exit(EXIT_FAILURE);
    }
    return err_loc;
}

std::vector<int> BMS_RS::rs_forney_syndromes(const std::vector<int>& synd, const std::vector<int>& pos, int nmess) {
    std::vector<int> erase_pos_reversed;
    for (auto p : pos) {
        erase_pos_reversed.push_back(nmess - 1 - p);
    }

    std::vector<int> fsynd(synd.begin() + 1, synd.end());

    for (int i = 0; i < pos.size(); i++) {
        int x = bms_gf.gf_pow(2, erase_pos_reversed[i]);
        for (int j = 0; j < fsynd.size() - 1; j++) {
            fsynd[j] = bms_gf.gf_mul(fsynd[j], x) ^ fsynd[j + 1];
        }
    }

    return fsynd;
}

std::vector<int> BMS_RS::rs_correct_errata(std::vector<int>& msg_in, std::vector<int>& synd, std::vector<int>& err_pos) {
    std::vector<int> coef_pos;
    for (int p : err_pos) {
        coef_pos.push_back(msg_in.size() - 1 - p);
    }

    std::vector<int> err_loc = rs_find_errata_locator(coef_pos);
    reverse(synd.begin(), synd.end());
    std::vector<int> err_eval = rs_find_error_evaluator(synd, err_loc, err_loc.size() - 1);

    std::vector<int> X;
    for (int i = 0; i < coef_pos.size(); i++) {
        int l = 255 - coef_pos[i];
        X.push_back(bms_gf.gf_pow(3, -l));
    }

    std::vector<int> E(msg_in.size(), 0);
    int Xlength = X.size();
    for (int i = 0; i < X.size(); i++) {
        int Xi_inv = bms_gf.gf_inverse(X[i]);

        std::vector<int> err_loc_prime_tmp;
        for (int j = 0; j < Xlength; j++) {
            if (j != i) {
                err_loc_prime_tmp.push_back(bms_gf.gf_sub(1, bms_gf.gf_mul(Xi_inv, X[j])));
            }
        }
        int err_loc_prime = 1;
        for (auto coef : err_loc_prime_tmp) {
            err_loc_prime = bms_gf.gf_mul(err_loc_prime, coef);
        }

        auto y = bms_gf.gf_poly_eval(err_eval, Xi_inv);
        y = bms_gf.gf_mul(bms_gf.gf_pow(X[i], 1 - 1), y);

        if (err_loc_prime == 0) {
            std::cerr << "ERROR: could not find magnitude" << std::endl;
            exit(EXIT_FAILURE);
        }

        auto magnitude = bms_gf.gf_div(y, err_loc_prime);
        E[err_pos[i]] = magnitude;
    }
    msg_in = bms_gf.gf_poly_add(msg_in, E);
    return msg_in;
}

std::vector<int> BMS_RS::rs_correct_msg(const std::vector<int>& msg_in, int nsym) {
    if (msg_in.size() > 255) {
        std::cerr << "ERROR: message too long" << std::endl;
        exit(EXIT_FAILURE);
    }

    std::vector<int> erase_pos;
    std::vector<int> msg_out = msg_in;

    if (erase_pos.size() > nsym) {
        std::cerr << "ERROR: too many errors" << std::endl;
        exit(EXIT_FAILURE);
    }

    auto synd = rs_calc_syndromes(msg_out, nsym);

    if (*std::max_element(synd.begin(), synd.end()) == 0) {
        // return {msg_out.begin(), msg_out.end() - nsym}, {msg_out.end() - nsym, msg_out.end()};
        return msg_out;
    }

    auto fsynd = rs_forney_syndromes(synd, erase_pos, msg_out.size());
    auto err_loc = rs_find_error_locator(fsynd, nsym, erase_pos.size());
    reverse(err_loc.begin(), err_loc.end());
    auto err_pos = rs_find_errors(err_loc, msg_out.size());

    if (err_pos.empty()) {
        std::cerr << "ERROR: error location failed" << std::endl;
        exit(EXIT_FAILURE);
    }

    std::vector<int> tmp = erase_pos;
    tmp.insert(tmp.end(), err_pos.begin(), err_pos.end());
    msg_out = rs_correct_errata(msg_out, synd, tmp);

    if (!rs_check(msg_out, nsym)) {
        std::cerr << "ERROR: message uncorrectable" << std::endl;
        exit(EXIT_FAILURE);
    }

    return msg_out;
}

/*
0110000101100010011000110111100001011111

011000100110110101110011001100000110110001111100

0100100001100101011011000110110001101111001011000010000001110111011011110111001001101100011001000010000110001101000100111111010011111001010000110001000011100101

0100001001100101011110100110010001110010011000010111010001101111011101100110010100100000011000010010000001101101011011110110001001101001011011000110111001101001001000000111001101101001011101000110010101000001100011101110100101101110011010100001110010011110

*/
